lib/ash_ai/mcp/server.ex

```diff
@@ -221,6 +222,92 @@ defmodule AshAi.Mcp.Server do
         # TODO: Cancel request?
         {:no_response, nil, session_id}
 
+      # TODO: this can support paginaton via params later
+      %{"method" => "resources/list", "id" => id} ->
+        resources =
+          opts
+          |> mcp_resources()
+          |> Enum.map(fn resource ->
+            %{
+              "name" => resource.name,
+              "description" => resource.description,
+              "uri" => resource.uri,
+              "title" => resource.title,
+              "mimeType" => resource.mime_type
+            }
+          end)
+
+        response = %{
+          "jsonrpc" => "2.0",
+          "id" => id,
+          "result" => %{
+            "resources" => resources
+          }
+        }
+
+        {:json_response, Jason.encode!(response), session_id}
+
+      %{"method" => "resources/read", "id" => id, "params" => %{"uri" => uri} = params} ->
+        opts =
+          opts
+          |> Keyword.update(
+            :context,
+            %{mcp_session_id: session_id},
+            &Map.put(&1, :mcp_session_id, session_id)
+          )
+
+        with %AshAi.McpResource{
+               uri: uri,
+               mime_type: mime_type
+             } = mcp_resource <-
+               find_mcp_resource_by_uri(uri, opts),
+             {:ok, text} <-
+               run_mcp_resource_action(mcp_resource, params, opts) do
+          response = %{
+            "jsonrpc" => "2.0",
+            "id" => id,
+            "result" => %{
+              "contents" => [
+                %{
+                  "uri" => uri,
+                  "mimeType" => mime_type,
+                  "text" => text
+                }
+              ]
+            }
+          }
+
+          {:json_response, Jason.encode!(response), session_id}
+        else
+          nil ->
+            response = %{
+              "jsonrpc" => "2.0",
+              "id" => id,
+              "error" => %{
+                "code" => -32_002,
+                "message" => "Resource not found",
+                "data" => %{"uri" => uri}
+              }
+            }
+
+            {:json_response, Jason.encode!(response), session_id}
+
+          {:error, error} ->
+            error_message = Exception.message(error)
```

Comment:
Usually, I would say we shouldn't put errors like this directly in the responses, but the tool/call also exposes the error as is , so I guess we should be ok